Add option to select replica movement strategy for rebalances

tomncooper opened the pull request:
### Type of change

Enhancement 

### Description

This adds the option to specify the `replicaMovementStrategies` for a cluster rebalance. By default `BaseReplicaMovementStrategy` is used, which randomly sorts the replica movements but other strategies [are available](https://github.com/linkedin/cruise-control/wiki/Pluggable-Components#replica-movement-strategy). The strategies can be specified in a comma separated list to form a priority chain. If a strategy earlier in the list finds two replica movements equivalent then these can be passed to the next strategy down to determine an order.

### Checklist

- [x] Make sure all tests pass
- [x] Update documentation
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Update CHANGELOG.md

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/KafkaRebalanceSpec.java:
```diff
@@ -41,6 +41,7 @@
     private int concurrentIntraBrokerPartitionMovements;
     private int concurrentLeaderMovements;
     private long replicationThrottle;
+    private String replicaMovementStrategies;
```
do you think we could use a `List<String>` ? the order should be preserved when serialized/deserialized right? or am I wrong?

tomncooper replied:
The order should be preserved. This is like the goals list in the `cruiseControl.spec`.

ppatierno replied:
So isn't it much better having a YAML array as we have for goals instead of a string with comma separated values? Just my 2 cents, but no need to change if you feel it's better in the current way. I just found YAML array a better fit for this field.

tomncooper replied:
We don't have a YAML array for the goals config in the Kafka CR, we have a single comma separated string split over several lines:
```
cruiseControl:
    capacity:
      networkIn: 10000KB/s
      networkOut: 10000KB/s
    config:
      default.goals: >
           com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,
           com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal,
           com.linkedin.kafka.cruisecontrol.analyzer.goals.DiskCapacityGoal
```

ppatierno commented:
I meant the goals in the KafkaRebalance resource, sorry

tomncooper commented:
I can make it an array for the `KafkaRebalance` resource. I suppose that would be consistent with the goals array in that resource. I would need to convert that back to a comma separated string to be sent to CC. WDYT?

ppatierno commented:
+1 from me :)

tomncooper commented:
It happens in the `CruiseControlParameters.asList(Iterable<String> values)` method. Which is called in the `PathBuilder.addParameter(CruiseControlParameters param, List<String> values)`.

ppatierno commented:
ok cool, thanks!

scholzj commented on api/src/main/java/io/strimzi/api/kafka/model/KafkaRebalanceSpec.java:
```diff
@@ -118,6 +119,15 @@ public void setReplicationThrottle(long bandwidth) {
         this.replicationThrottle = bandwidth;
     }
 
+    @Description("A list of strategies used to determine the execution order for the generated optimization proposal. By default BaseReplicaMovementStrategy is used, which is totally random.")
```
Not a native speaker ... but _totally random_ sounds very informal for documentation. I also doubt it is really random - i.e. randomizing the list of movements ... I guess it might be just as it is generated by some algo or something like that I guess. Wouldn't something like `By default BaseReplicaMovementStrategy is used, which moves the replicas without any specific order.` be a bit better?

samuel-hawker replied:
Agreed with Jakub here, his wording looks good to me, or if it is truly random (or atleast different with the same input) - then it might be worth saying it is _non-deterministic_

tomncooper replied:
Sorry, this is copy of CC's wording, which I removed from other docstrings but didn't change here. I will change it.

tomncooper commented:
The default strategies can currently be set using the `Kafka.spec.cruiseControl.spec.default.replica.movement.strategies` key. But they would be, like the current default goals, set as a comma separated string. That is because it is just a pass through of a CC config value via the additional properties catch all. Would it be worth, for consistency, making proper entries for `default.goals` and `default.replica.movement.strategies` in the Kafka CRD which are YAML arrays?

scholzj commented:
> The default strategies can currently be set using the Kafka.spec.cruiseControl.spec.default.replica.movement.strategies key. But they would be, like the current default goals, set as a comma separated string. That is because it is just a pass through of a CC config value via the additional properties catch all. Would it be worth, for consistency, making proper entries for default.goals and default.replica.movement.strategies in the Kafka CRD which are YAML arrays?

@tomncooper To be honest, I think I would let you, @kyguy and @ppatierno decide on this one. I can live with both.

kyguy commented:
> @tomncooper To be honest, I think I would let you, @kyguy and @ppatierno decide on this one. I can live with both

PR looks great! As discussed offline, we can leave these changes for another PR should we decide to do this!

tomncooper commented:
I agree, I think leaving the CC server level configs are they are (comma separated lists) make sense as they are currently passed straight through to CC in a config map.

scholzj commented:
@strimzi-ci run tests profile=cruisecontrol

tomncooper commented:
I updated the rebalance performance tuning docs. Which TBF I should have done from the start. @laidan6000 could you have a look please. I also changed the formatting of the table to make it less wide.

scholzj commented:
@tomncooper The review would be much easier if you stop squashing the commits every time. We would be able to just check the new commits and not the whole PR again and again.

tomncooper commented on documentation/modules/cruise-control/con-rebalance-performance.adoc:
```diff
@@ -17,32 +19,56 @@ These commands consist of two types of operations:
+| `default.replica.movement.strategies`             .2+| 
+ The list of strategies (in priority order) used to determine the execution order for generated proposals. A comma separated string for server and a YAML array for KafkaRebalance. 
+.2+| `BaseReplicaMovementStrategy`
+| `replicaMovementStrategies`
```
They are listed in the section above so I think it would be redundant to list them again here.

tomncooper commented:
@strimzi-ci run tests profile=cruisecontrol
